File: src/Bench.go
package Bench

import (
	"fmt"
	"sync/atomic"
	"time"
)

var benchmarkEpoch = time.Now()

func BenchNow() float64 {
	return float64(time.Since(benchmarkEpoch).Nanoseconds()) / 1e3
}

func FormatNumber(n float64) string {
	return fmt.Sprintf("%.6f", n)
}

func Opaque(a interface{}) func() interface{} {
	return func() interface{} {
		return opaqueValue(a)
	}
}

// Keep benchmark inputs opaque to the Go compiler without constraining kernels.
//go:noinline
func opaqueValue(a interface{}) interface{} {
	return a
}

var benchmarkResult int64

func MeasureBatch(iterations int, expected int, act func() int) float64 {
	result := 0
	start := time.Now()
	for i := 0; i < iterations; i++ {
		result = act()
		atomic.StoreInt64(&benchmarkResult, int64(result))
	}
	elapsed := float64(time.Since(start).Nanoseconds()) / 1e3
	if result != expected {
		panic("Unstable benchmark result")
	}
	return elapsed
}

File: src/Bench.erl
-module(bench@foreign).
-export([benchNow/0, opaque/1, formatNumber/1, measureBatch/3]).

benchNow() -> fun() ->
  erlang:convert_time_unit(erlang:monotonic_time() - erlang:system_info(start_time),
                          native, nanosecond) / 1000.0
end.
opaque(A) -> fun() -> A end.
formatNumber(N) -> erlang:float_to_binary(N, [{decimals, 6}]).
measureBatch(Iterations, Expected, Act) -> fun() ->
  Start = erlang:monotonic_time(),
  Result = measure_loop(Iterations, Act, 0),
  Elapsed = erlang:convert_time_unit(erlang:monotonic_time() - Start, native, nanosecond) / 1000.0,
  case Result =:= Expected of
    true -> Elapsed / Iterations;
    false -> erlang:error({unstable_benchmark_result, Result, Expected})
  end
end.
measure_loop(0, _Act, Result) -> Result;
measure_loop(N, Act, _Previous) ->
  Result = Act(),
  put(benchmark_result, Result),
  measure_loop(N - 1, Act, Result).
